authenticator/nonce_validator.go

```diff
@@ -0,0 +1,63 @@
+package authenticator
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"time"
+)
+
+var defaultMaxTimeGap = time.Second * 30
+
+//ErrNonceNotInRange error when nonce not in acceptable range.
+var ErrNonceNotInRange = errors.New("submitted nonce is not in acceptable range")
+
+// NonceValidator checking validate by time range
+type NonceValidator struct {
+	// MaxTimeGap is max time different between client submit timestamp
+	// and server time that considered valid. The time precision is millisecond.
+	MaxTimeGap time.Duration
+}
+
+// NonceValidatorOption is the option of Nonce Validator constructor.
+type NonceValidatorOption func(*NonceValidator)
+
+// NonceValidatorWithMaxTimeGap is the option to create NonceValidator with custom time gap.
+func NonceValidatorWithMaxTimeGap(gap time.Duration) NonceValidatorOption {
+	return func(validator *NonceValidator) {
+		validator.MaxTimeGap = gap
+	}
+}
+
+// NewNonceValidator return NonceValidator with default value (30 second)
+func NewNonceValidator(options ...NonceValidatorOption) *NonceValidator {
+	v := &NonceValidator{
```

Comment:
Can we just define object with default value like v := &NonceValidator{}? cause in line 41 if an NonceValidatorOption change MaxTimeGap to zero we have to assign MaxTimeGap = defaultMaxTimeGap twice.

Reply:
yes, we should, good spot.

Reply:
did that in 724a846c3429c0697e8f92110f71aa5f226b88dc.